app/models/queued_post.rb

```diff
@@ -60,24 +64,23 @@ def create_options
   end
 
   def approve!(approved_by)
-    created_post = nil
+    created_post = PostCreator.new(user, create_options.merge(skip_validations: true, acting_user: approved_by)).create!
 
-    creator = PostCreator.new(user, create_options.merge(skip_validations: true, skip_jobs: true))
-    QueuedPost.transaction do
-      change_to!(:approved, approved_by)
+    created_post.revise(
+      User.find(post_options['changes']['editor_id']),
+      post_options['changes'],
+      skip_validations: true,
+      force_new_version: true
+    ) if revised?
 
+    transaction do
       UserBlocker.unblock(user, approved_by) if user.blocked?
-
-      created_post = creator.create
-
-      unless created_post && creator.errors.blank?
-        raise StandardError.new(creator.errors.full_messages.join(" "))
-      end
+      change_to!(:approved, approved_by)
     end
-
-    # Do sidekiq work outside of the transaction
-    creator.enqueue_jobs
-
+  rescue
+    PostDestroyer.new(approved_by, created_post).destroy if created_post&.persisted?
+    raise
+  else
```

Comment:
It's critical to place the transaction block at the end, so that

- should anything raise during post creation/revision: the created post is destroyed.
- should anything raise inside the transaction block: the transaction is rolled back and the created post is destroyed.